### ISSUE TITLE
Implement "your role has been changed" message

### ISSUE THREAD

**Issue by kslottow** [pculture/unisubs#3397]
- Recipient: team member whose role has been changed
- [Copy templates](https://docs.google.com/document/d/18C-tmHGlRBI-yBTwr23AMZ73u0C9_46ICFxcT_qfOXQ/edit) - this notification is 1.6 on the list
- See #3434 for how we're implemented these new notifications in general

**Comment by bendk** [comment]
As part of this ticket I'm updating the code to handle the change member role form.  Please make sure to test that I didn't break anything.

**Comment by bendk** [comment]
Implemented in the gh-3397 branch

**Comment by Rachel36** [comment]
Hi @bendk! I changed a member's role from a manager to an admin and did not receive a message as I expected to.  Here are the steps to reproduce: 

- [ ] Logged in as a superuser, changed members role- admin/teams/teammember

- [ ] Logged in to app as team member who's role had been changed

- [ ]  Checked message box, no messages available. 

*Note: I did make sure to run migrations. Including a pic for your reference and to make sure I did this correctly. 

![image](https://user-images.githubusercontent.com/42118332/46307304-2b6d4c80-c57c-11e8-8b32-0e116f2a6973.png)

**Comment by PCF-Testing** [comment]
@Rachel36:

> Logged in as a superuser, changed members role- admin/teams/teammember

You used Django admin to change the role, which is why the message was not sent. Use of Django admin is not a legitimate way of operating the system; rather, it is a way to hack the data when something has got messed up or a bug in the front-ed UI makes some things temporarily non-functional. You cannot expect a change in the admin to do everything - send the related messages, update the database indexes - the same way a change made through the regular front-end UI does, so I wouldn't even bother to test this. Just make sure things work well when the change is made through the front-end UI or through API, that should be enough.

**Comment by Rachel36** [comment]
Thanks for your help @bendk and @PCF-Testing! I ran a test with a member who has both notify by email and notify by message set. Changed the role through regular UI. Still did not receive a message or email. So it appears as though at least message isn't coming through? I chatted with Maggie this morning, working through my settings for sendgrid to verify whether or not email came through.

**Comment by bendk** [comment]
If notify by email is set, then the message won't be created.  For these new notifications, notify by message is just ignored.

If you want to see the message, disable notify by email.

If you want to see the email, you need to find the text output in the console that you're running `dev up` in.

**Comment by PCF-Testing** [comment]
> If notify by email is set, then the message won't be created. For these new notifications, notify by message is just ignored.

@bendk: curiously, why is it done this way? So far, we have been duplicating all messages both by email and by internal IM, and it might be confusing not to see copies of emails in the onsite inbox. Also, email messages are less reliable because they can get blocked by sendgrid or land in the spam box on the user's end; if copies don't get sent as internal messages, the chances of discovering that something is wrong decrease.

**Comment by bendk** [comment]
I think the main reason is that it's redundant to have an email and a message that says the same thing.  If a user sees the notification in their email, then they got to the site and see the exact same message then that seems like it could be annoying.  Users also might learn to just ignore the "you have XX new messages" announcement, since they know they've probably already seen them.

That said, we can definitely think about it more.  You're point about emails being less reliable makes a lot of sense.

Maybe we could tie them together somehow.  So if the user opens the email, or maybe if they click a link in the email, then we mark the message as read.  That would require some extra code though.

**Comment by bendk** [comment]
> ￼For current program/language manager: changed project setting to "none" and selected a different language. (Please see email below)

I think something got messed up with the github formatting.  Is there an issue to fix here?

> Should I receive a notification if role/project assignment remains the same, but languages are altered? Reading the "New Style Teams Notifications Copy Templates" doc, I expected a notification re: languages which are now set? Did not receive notification when I only made changes to languages.

I got a little confused trying to follow this one, but if you use the edit role dialog to and keep a user as a language/project manager and only change the languages, then the user should get a message.  From my testing this is happening though.  Can you check this one out more?  If you are still seeing it, please give more details.

**Comment by Rachel36** [comment]
@bendk - changes look good! I'm now receiving all notifications as I expect to. I only have one last question before verifying this one - should a user receive notification if they are removed from a team? In my testing this did not occur. Just want to confirm that is correct?

**Comment by bendk** [comment]
@Rachel36 that's a different issue.  I think they should, but we'll implement that later.

**Comment by Rachel36** [comment]
Thanks @bendk! Enterprise-3397 has been verified.

**Comment by bendk** [comment]
@PCF-Testing @Rachel36 After talking with Kenzie on Friday, we're thinking that we should try to match the old behavior as much as possible -- at least at first.  In my mind this means:

  - Making it so we use the notify_by_message flag to determine if we send messages or not.
    - If this is checked, then we will always send a message, even if we also send an email
    - If this is unchecked, then we will never send a message, even for messages like `new active assignment` where we always send an email.

I'm hoping that this matches the most typical behavior of the current messaging code.

**Comment by Rachel36** [comment]
Hi @bendk!
 
> I'm hoping that this matches the most typical behavior of the current messaging code.

Does that mean there will be changes to this issue that I should test, or another issue created?  Throughout testing on 3397, if the email flag was set, then a message was not created.

**Comment by bendk** [comment]
@Rachel36 yes, can you test that part?  I'm pushing the code right now.  Sorry for the extra work here.

**Comment by Rachel36** [comment]
@bendk - no worries! I'll start testing now!

**Comment by Rachel36** [comment]
@bendk All testing looks great, with only one exception: 

> If this is unchecked, then we will never send a message, even for messages like new active assignment where we always send an email.

I did not receive an email regarding a new active assignment. I expected to receive this message after assigning a subtitle request to a team member. Is this the correct scenario in which a "new active assignment" notification would be sent?  If so, I did not receive an email when the message flag was unchecked -  whether or not the email flag was set.

**Comment by bendk** [comment]
For dev instances "send an email" means print it out to the console.  Did you check there?

**Comment by Rachel36** [comment]
@bendk Yes sir. No luck.

**Comment by bendk** [comment]
Are you using this branch to test the active assignment code?  That's implemented in gh-enterprise-1894.

**Comment by Rachel36** [comment]
@bendk 3397 has been verified.  Will check the "new active assignment" message under gh-1894.

**Comment by bendk** [comment]
The templates say it should say "You've been promoted" in the case where the change is a promotion.  I interpreted that as meaning:
 - The role changes to admin/manager from a "lower" role
 - The role stayed as contributor, before the user didn't have any project/language managed now they have at least 1

**Comment by PCF-Testing** [comment]
Deployed.
